optimade/filtertransformers/mongo.py

```diff
@@ -208,30 +219,85 @@ def property_zip_addon(self, arg):
+            expr2 = arg["$and"][1]
+            key = list(expr1.keys())[0]
+            if expr1.keys() == expr2.keys():
+                if expr1.get(key) == {"$ne": None}:
+                    return self._recursive_expression_phrase(["NOT", expr2])
+                elif expr2.get(key) == {"$ne": None}:
+                    return self._recursive_expression_phrase(["NOT", expr1])
+            return {
+                "$or": [
+                    self._recursive_expression_phrase(["NOT", subdict])
+                    for subdict in arg["$and"]
+                ]
+            }
+
+        def handle_not_or(arg):
+            """Handle the case of ~(A | B) -> (~A & ~B).
+
+            !!! note 
+            Although the MongoDB `$nor` could be used here, it is not convenient as it
+            will also return documents where the filtered field is missing when testing
+            for inequality.
+
+            """
 
-        # handle the case of {"$not": {"$and": [expr1, expr2]}} using per-expression negation,
-        # e.g. {"$and": [{prop1: {"$not": expr1}}, {prop2: {"$not": ~expr2}}]}.
-        # Note that this is not the same as NOT (expr1 AND expr2)!
-        if "$and" in arg[1]:
             return {
                 "$and": [
                     self._recursive_expression_phrase(["NOT", subdict])
-                    for subdict in arg[1]["$and"]
+                    for subdict in arg["$or"]
                 ]
             }
 
-        # simple case of negating one expression, from NOT (expr) to ~expr.
-        return {prop: {"$not": expr} for prop, expr in arg[1].items()}
+        if len(arg) == 1:
+            # without NOT
+            return arg[0]
+
+        if "$or" in arg[1]:
+            return handle_not_or(arg[1])
+
+        if "$and" in arg[1]:
+            return handle_not_and(arg[1])
+
+        # If the NOT operator occurs at the lowest nesting level,
+        # the expression can be simplified by using the opposite operator and removing the not.
+        for prop, expr in arg[1].items():
+            for operator, value in expr.items():
+                if operator in self.inverse_operator_map:
+                    if operator in ("$in", "$eq"):
+                        return {
+                            "$and": [
+                                {
+                                    prop: {
+                                        self.inverse_operator_map.get(operator): value
+                                    }
+                                },
+                                {prop: {"$ne": None}},
+                            ]
+                        }
+
+                    else:
+                        return {prop: {self.inverse_operator_map.get(operator): value}}
+                else:
+                    if "#known" in expr:
+                        return {prop: {"$not": expr}}
+                    else:
+                        return {"$and": [{prop: {"$not": expr}}, {prop: {"$ne": None}}]}
```

Comment:
I'm still just trying to work through this bit

Reply:
Rather than all of this nesting, I'd prefer to be explicit about the expectation of what `arg[1]` is at this point, something like:

```suggestion
        prop, expr = next(iter(arg[1].items()))
        operator, value = next(iter(expr.items()))
        if operator in self.inverse_operator_map:
            filter_ = {prop: {self.inverse_operator_map[operator]: value}}
            if operator in ("$in", "$eq"):
                filter_ = {"$and": [filter_, {prop: {"$ne": None}}]}
            return filter_
            
        filter_ = {prop: {"$not": expr}}
        if "#known" in expr:
            return filter_
        return {"$and": [filter_, {prop: {"$ne": None}}]}
```

Reply:
Thanks for your help cleaning up this piece of code. I think the "for" construction I used to extract the values from the dictionary occurs in more often in the optimade code. Perhaps it is worthwhile to clean this up elsewhere in the code as well.